Frameworks/Foundation/NSData.mm

```diff
@@ -303,69 +305,59 @@ - (instancetype)initWithContentsOfFile:(NSString*)filename {
-    char* fname = (char*)[filename UTF8String];
-
-    TraceVerbose(TAG, L"NSData writing %hs (%d bytes)", fname, [self length]);
-    if (!fname) {
-        TraceVerbose(TAG, L"Filename is null!");
-        return FALSE;
-    }
-    EbrFile* fpOut = EbrFopen((const char*)fname, "wb");
-    if (fpOut) {
-        EbrFwrite([self bytes], 1, [self length], fpOut);
-        EbrFclose(fpOut);
-
-        return TRUE;
-    } else {
-        TraceVerbose(TAG, L"NSData couldn't open %hs for write", fname);
-        return FALSE;
-    }
+    return [self writeToURL:[NSURL fileURLWithPath:filename] atomically:atomically];
 }
 
 /**
  @Status Caveat
  @Notes Only file:// URLs supported. atomically parameter not supported.
 */
 - (BOOL)writeToURL:(NSURL*)url atomically:(BOOL)atomically {
-    if (![url isFileURL]) {
-        TraceVerbose(TAG, L"-[NSData::writeToURL]: Only file: URLs are supported. (%hs)", [[url absoluteString] UTF8String]);
-        return NO;
-    }
-    return [self writeToFile:[url path] atomically:atomically];
+    return [self writeToURL:url options:(atomically ? NSDataWritingAtomic : 0) error:nil];
 }
 
 /**
  @Status Caveat
  @Notes options parameter not supported
 */
 - (BOOL)writeToFile:(NSString*)filename options:(NSDataWritingOptions)options error:(NSError**)error {
-    char* fname = (char*)[filename UTF8String];
-
-    TraceVerbose(TAG, L"NSData writing %hs (%d bytes)", fname, [self length]);
-    EbrFile* fpOut = EbrFopen((const char*)fname, "wb");
-    if (fpOut) {
-        EbrFwrite([self bytes], 1, [self length], fpOut);
-        EbrFclose(fpOut);
-        return YES;
-    }
-
-    TraceVerbose(TAG, L"NSData couldn't open %hs for write (with options)", fname);
-    if (error) {
-        *error = [NSError errorWithDomain:NSCocoaErrorDomain code:NSFileNoSuchFileError userInfo:@{ NSFilePathErrorKey : filename }];
-    }
-    return NO;
+    return [self writeToURL:[NSURL fileURLWithPath:filename] options:options error:error];
 }
 
 /**
  @Status Caveat
  @Notes options parameter not supported
 */
 - (BOOL)writeToURL:(NSURL*)url options:(NSDataWritingOptions)options error:(NSError**)errorp {
-    if (![url isFileURL]) {
-        TraceVerbose(TAG, L"-[NSData::writeToURL]: Only file: URLs are supported. (%hs)", [[url absoluteString] UTF8String]);
+    if (!url) {
+        if (errorp) {
+            *errorp = [NSError errorWithDomain:@"NSData" code:100 userInfo:nil];
+        }
+        return NO;
+    }
+
+    NSOutputStream* outputStream =
+        [NSOutputStream outputStreamWithURL:url append:(0 != (options & NSDataWritingWithoutOverwriting)) ? YES : NO];
+    [outputStream open];
+
+    if (NSStreamStatusOpen != outputStream.streamStatus) {
+        if (errorp) {
+            *errorp = [NSError errorWithDomain:@"NSData" code:100 userInfo:nil];
+        }
+        return NO;
+    }
+    auto numBytes = [self length];
+    auto result = [outputStream write:(const unsigned char*)[self bytes] maxLength:numBytes];
```

Comment:
> result = [outputStream write:(const unsigned char*)[self bytes] maxLength:numBytes]; [](start = 8, length = 85)

we might not write the whole buffer, we need to check if we have any left over and write the rest?
I recall similar code existing already.